app/digitizingcontroller.cpp

```diff
@@ -70,6 +70,12 @@ bool DigitizingController::hasPointGeometry(QgsVectorLayer *layer) const
   return layer && layer->geometryType() == QgsWkbTypes::PointGeometry;
 }
 
+// TODO move second condition to QgsQuickFeatureLayerPair class
+bool DigitizingController::isPairValid(QgsQuickFeatureLayerPair pair) const
+{
+    return pair.isValid() && pair.feature().geometry().isGeosValid();
```

Comment:
Id like to add validity test which restricts cases when polygon/polyline doesn't have enough points. It's called from QML, where if its false, a notification message is shown. Without this check, feature form is opened in such cases which imho its not correct. However, I would like tomove it to QgsQuickFeatureLayerPair::isValid function. What do you think?

Reply:
Ok in that case better to check vertex count (>= 2 for lines, >= 3 for polygons) instead of GEOS validity. For example a polygon with self-intersection is not valid according to GEOS but we should allow adding such geometry.

I would probably keep the function here as it is specific to digitizing... unless we also need it elsewhere?

Reply:
There is another problem with self-intersection of polygons regarding highlight. Although at the end such polygon can be created and its rendered on canvas, currently its not visible while recording since any polygon becomes self-intersectioned:
`polygon rings self-intersect or intersect each other - skipping`

Reply:
I know - that's something we should fix at some point within the feature highlight object...